panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,732 @@
+                                             TRUE, /* save to disk */
+                                             NULL,
+                                             on_network_changes_committed_cb,
+                                             NULL);
+}
+
+static void
+get_active_connection_and_device (CcUpdatesPanel  *self,
+                                  NMDevice       **out_device,
+                                  NMConnection   **out_connection,
+                                  NMAccessPoint  **out_ap)
+{
+  NMActiveConnection *active_connection = NULL;
+  NMConnection *connection = NULL;
+  const GPtrArray *active_devices = NULL;
+  NMAccessPoint *ap = NULL;
+  NMDevice *active_device = NULL;
+
+  active_connection = nm_client_get_primary_connection (self->nm_client);
+
+  /* If no primary connection is already present, try and use the connecting one */
+  if (!active_connection)
+    active_connection = nm_client_get_activating_connection (self->nm_client);
+
+  if (active_connection)
+    {
+      connection = NM_CONNECTION (nm_active_connection_get_connection (active_connection));
+      active_devices = nm_active_connection_get_devices (active_connection);
+
+      if (active_devices && active_devices->len > 0)
+        {
+          /* This array is guaranteed to have only one element */
+          active_device = g_ptr_array_index (active_devices, 0);
+
+          if (NM_IS_DEVICE_WIFI (active_device))
+            ap = nm_device_wifi_get_active_access_point (NM_DEVICE_WIFI (active_device));
+        }
+    }
+
+  if (out_device)
+    *out_device = active_device;
+
+  if (out_connection)
+    *out_connection = connection;
+
+  if (out_ap)
+    *out_ap = ap;
+}
+
+static NMSettingConnection *
+ensure_setting_connection (NMConnection *connection)
+{
+  NMSettingConnection *setting = nm_connection_get_setting_connection (connection);
+
+  g_assert (connection != NULL);
+
+  if (!setting)
+    {
+      setting = NM_SETTING_CONNECTION (nm_setting_connection_new ());
+      nm_connection_add_setting (connection, NM_SETTING (setting));
+    }
+
+  return setting;
+}
+
+static NMSettingUser *
+ensure_setting_user (CcUpdatesPanel *self,
+                     NMConnection   *connection)
+{
+  NMSettingUser *setting_user;
+
+  g_assert (connection != NULL);
+
+  setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+
+  if (!setting_user)
+    {
+      g_autoptr(GError) error = NULL;
+      NMSettingConnection *setting;
+      NMMetered metered;
+      gboolean enabled;
+
+      g_debug ("Creating a new custom config file for the current connection…");
+
+      /* Add a new NMSettingUser to this connection */
+      setting_user = NM_SETTING_USER (nm_setting_user_new ());
+      nm_connection_add_setting (connection, NM_SETTING (setting_user));
+
+      /* The default value depends on the metered state of the connection */
```

Comment:
I was thinking that `ensure_setting_user()` would just be the lines up to `nm_connection_add_setting()` (i.e. like `ensure_setting_connection()`). The lines below here which set up signal handlers and modify the UI should probably be moved elsewhere — probably to just after where you first call `ensure_setting_user()`.

Reply:
I disagree. The new panel assumes that we always have a value set for those settings. I don't see what's the problem with those lines. I also see no point in splitting this part of code, for it is too related to what's happening there. If no NMSettingUser is set, set the default and save and return; otherwise, just return.

There are no signal handlers being set.

Reply:
Also, the whole point of the assert there is that it is absolutely wrong to reach this function if there's no NMSettingUser set; this function **cannot** set a NMSettingUser.

Reply:
I just reworked the code as I had in mind, and I think it’s an improvement. See what you think [here](https://github.com/endlessm/gnome-control-center/commit/8386a9170525f0326fbfae7cd2e2e9faa7537c29). Hopefully it explains my idea a bit better.

Another reason I prefer the approach I’m suggesting is that it means you don’t end up always storing the default values on the `NMSettingUser` (only if the user changes the values to something, then back to the defaults). It seems a bit wasteful to create an `NMSettingUser`, then immediately set the default values on it, given that you were already handling the lack of a return value from the `nm_setting_user_get_value()` calls.